lib-php/Hierarchy.php

```diff
@@ -0,0 +1,148 @@
+        if ($iAdminLevel > 11) {
+            return [];
+        }
+
+        // For the immediate children of the country (admin_level=2 is always
+        // the country) we need a special query, because there won't be any
+        // records in `place_addressline`
+        if ($iAdminLevel == 3) {
+            $sSQL = "
+                SELECT
+                    p2.osm_type,
+                    p2.osm_id,
+                    COALESCE(
+                        p2.name -> 'int_name',
+                        p2.name -> 'alt_name',
+                        p2.name -> 'name'
+                    ) AS name,
+                    TO_CHAR(
+                        TO_TIMESTAMP(EXTRACT(epoch FROM p2.indexed_date)),
+                        'YYYY-MM-DD\"T\"HH:MI:SS+00:00'
+                    ) AS indexed_date
+                FROM placex AS p2
+                WHERE p2.country_code = :countryCode
+                AND p2.class = 'boundary'
+                AND p2.type = 'administrative'
+                AND p2.osm_type = 'R'
+                AND admin_level = (
+                    SELECT DISTINCT p1.admin_level
+                    FROM placex AS p1
+                    WHERE p1.country_code = :countryCode
+                    AND p1.class = 'boundary'
+                    AND p1.type = 'administrative'
+                    AND p1.osm_type = 'R'
+                    AND p1.admin_level > CASE
+                        WHEN p1.country_code = 'mt' THEN 3
+                        ELSE 2
+                        END
+                    ORDER BY p1.admin_level ASC
+                    LIMIT 1
+                )
+            ";
+
+            $aResult = $this->oDB->getAll($sSQL, [
+                ':countryCode' => $sCountryCode,
+            ]);
+        } else {
+            $sSQL = "
+                SELECT
+                    p2.osm_type,
+                    p2.osm_id,
+                    COALESCE(
+                        p2.name -> 'int_name',
+                        p2.name -> 'alt_name',
+                        p2.name -> 'name'
+                    ) AS name,
+                    TO_CHAR(
+                        TO_TIMESTAMP(EXTRACT(epoch FROM p2.indexed_date)),
+                        'YYYY-MM-DD\"T\"HH:MI:SS+00:00'
+                    ) AS indexed_date
+                FROM placex AS p1, placex AS p2, place_addressline AS pa
+                WHERE p1.osm_id = :osmId
+                AND p1.osm_type = 'R'
+                AND pa.address_place_id = p1.place_id
+                AND pa.place_id = p2.place_id
+                AND pa.isaddress
+                AND pa.fromarea
+                AND p2.admin_level = :adminLevel
+                AND p2.class = 'boundary'
+                AND p2.type = 'administrative'
+                AND p2.osm_type = 'R'
+            ";
+
+            $aResult = $this->oDB->getAll($sSQL, [
+                ':osmId' => $iOsmId,
+                ':adminLevel' => $iAdminLevel,
+            ]);
+        }
```

Comment:
Ich würde gerne diese beiden Abfragen in eine tun. Weiß aber nicht wie...

Reply:
die beiden Queries sind bis auf die SELECT clause ziemlich unterschiedlich. Diese könnte man in nen extra String packen und dann zusammen`.`fügen, ansonsten sehe ich keinen echten Mehrgewinn.